modules/index-rebuild/src/main/java/org/opencastproject/index/rebuild/IndexRebuildService.java

```diff
@@ -0,0 +1,228 @@
+    recreateService(index, service);
+  }
+
+  /**
+   * Trigger repopulation of the specified index with data from the specified service.
+   *
+   * @param index
+   *           The index to rebuild.
+   * @param service
+   *          The {@link IndexRebuildService.Service} to re-add data from.
+   *
+   * @throws IndexRebuildException
+   *           Thrown if the index rebuild failed.
+   */
+  private void recreateService(AbstractSearchIndex index, IndexRebuildService.Service service)
+          throws IndexRebuildException {
+
+    if (!indexProducers.containsKey(service)) {
+      throw new IllegalStateException(format("Service %s is not available", service));
+    }
+
+    IndexProducer indexProducer = indexProducers.get(service);
+    logger.info("Starting to recreate index {} for service '{}'", index.getIndexName(), service);
+    indexProducer.repopulate(index.getIndexName());
+    logger.info("Finished to recreate index {} for service '{}'", index.getIndexName(), service);
+  }
+
+  /**
+   * Add IndexProducer to internal map.
+   *
+   * @param indexProducer
+   *           The IndexProducer to add.
+   */
+  private void addIndexProducer(IndexProducer indexProducer) {
+    indexProducers.put(indexProducer.getService(), indexProducer);
+    logger.info("Service {} added.", indexProducer.getService());
+  }
+
+  /**
+   * Remove IndexProducer from internal map.
+   *
+   * @param indexProducer
+   *           The IndexProducer to remove.
+   */
+  private void removeIndexProducer(IndexProducer indexProducer) {
+    indexProducers.remove(indexProducer.getService(), indexProducer);
+    logger.info("Service {} removed.", indexProducer.getService());
+  }
+
+  /**
+   * Unregister this service at OSGI.
+   */
+  private void unregisterIndexRebuildService() {
+    if (serviceRegistration != null)  {
+      logger.info("Unregister IndexRebuildService.");
+      serviceRegistration.unregister();
+      serviceRegistration = null;
+    }
+  }
+
+  /**
+   * Register this service at OSGI.
+   *
+   * @param bundleContext
+   *           The bundle context.
+   */
+  private void registerIndexRebuildService(BundleContext bundleContext) {
+    if (serviceRegistration == null) {
+      logger.info("Register IndexRebuildService.");
+      serviceRegistration = bundleContext.registerService(this.getClass().getName(), IndexRebuildService.this, null);
+    }
+  }
+
+  /**
+   * Listen to changes in the availability of IndexProducer services.
+   */
+  private final class IndexProducerListener implements ServiceListener {
+
+    private final BundleContext bundleContext;
+
+    /**
+     * Constructor.
+     *
+     * @param bundleContext
+     *           The bundle context.
+     */
+    private IndexProducerListener(BundleContext bundleContext) {
+      this.bundleContext = bundleContext;
+    }
+
+    @Override
+    public void serviceChanged(ServiceEvent serviceEvent) {
```

Comment:
If I interpret this correctly, you are manually searching for OSGi services of a given type. Why not simply use a dynamic binding?

Reply:
Since I couldn't figure out how to only activate the IndexRebuildService if we have one service of each type obtained. (We can't just skip a service during the rebuild, since the order is important...) Maybe I missed something? I know I can set MinimumCardinality, but that's a hardcoded number...

Reply:
Couldn't you just have something like this in `activate()`:

```pascal
while services.length() < X; do
  sleep(1);
end
```

Or just refuse to start if you don't have all the services.
It's similar to what you do here just with less confusing custom code.

Reply:
Okay, I could've done that, but I think then I prefer my solution. I'm not happy about the additional code either, but the basic principle isn't that hard to understand, and I think it's a much cleaner solution to use what OSGI has to offer us than busy waiting or failing later.

Reply:
I agree that the basic principle isn't that hard but as you said, you had to read through a lot of OSGi documentation to actually get what all components and methods actually do. This means that understanding this and figuring out any problems will probably not be easy for people while not adding much benefit.

So, my thoughts are basically that I would probably be annoyed if I had to deal with this code in a year or two since this is doing “something special” and not doing what basically everybody else does. Not sure if you would like to deal with this code in two years.

---

Maybe another option would also be to use static bindings:

Right now (with this patch applied), to add a new index you would

- Implement the Interface `IndexProducer`
- Add a new type to `IndexRebuildService.Service` in this class
- Make `getService()` return this type

So, you have to implement the service and modify thgis class anyway.
Instead, we could just use a static binding which would then really simplify the OSGi magic.
That would mean that implementing a new service would be:

- Implement the Interface `IndexProducer`
- Add binding for the specific new implementation to `IndexRebuildService`

---

Or yet another idea, just use a dynamic binding and return an error when starting the rebuild if not all service types are available.

---

In the end, this is no veto against getting this in. It should work like this and is a clever approach. I just wouldn't want to deal with this later. But if others do, I'm fine with it ;-p

Reply:
Lars and I talked about this some more and realized that the alternatives aren't really that much better. I don't like just returning an error if services aren't available, but having a static binding for each service isn't that trivial either since we can't reference the sub classes without introducing cyclic dependencies (since all these sub classes implement IndexProducer from the same module), so we would need to differentiate them using target properties, and those are essentially hard-coded strings... so we came to the agreement that I would add a comment explaining how the listening stuff works to make this a bit more maintainable.